common/logger/src/lib.rs

```diff
@@ -1,74 +1,426 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-#![forbid(unsafe_code)]
-
-use chrono::Utc;
+pub use chrono::Utc;
 use env_logger::filter;
 pub use log::{self, Level, Log, Metadata, Record};
+pub use serde_json;
+
+use serde::Serialize;
+use serde_json::Value;
 use std::{
+    collections::HashMap,
     env,
     fmt::{self, Write},
-    sync::mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    fs::{File, OpenOptions},
+    io,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    },
     thread,
 };
 
 pub mod prelude {
     pub use crate::{crit, debug, error, info, trace, warn};
 }
+pub use std::io::Write as IoWrite;
 
 pub const CHANNEL_SIZE: usize = 256;
 pub const DEFAULT_TARGET: &str = "libra";
 const RUST_LOG: &str = "RUST_LOG";
 
+pub trait StructLogSink: Sync {
+    fn send(&self, entry: StructuredLogEntry);
+}
+
+// This is poor's man AtomicReference from crossbeam
+// It have few unsafe lines, but does not require extra dependency
+static NOP: NopStructLog = NopStructLog {};
+static mut STRUCT_LOGGER: &'static dyn StructLogSink = &NOP;
+static STRUCT_LOGGER_STATE: AtomicUsize = AtomicUsize::new(UNINITIALIZED);
+const UNINITIALIZED: usize = 0;
+
+const INITIALIZING: usize = 1;
+const INITIALIZED: usize = 2;
+
+#[derive(Default, Serialize)]
+pub struct StructuredLogEntry {
+    #[serde(skip_serializing_if = "Option::is_none")]
```

Comment:
what is the behavior in JSON if the option is None? At least in TOML it doesn't serialize the line, so these macros would be redundant if we were using TOML.

Reply:
Default behavior puts None into json